Single node DDP: "Default process group is not initialized"

## 🐛 Bug
Unable to start single node ddp training on 0.8.0

### To Reproduce
~~was going to run the gpu_template but... #2235~~
both methods of running the template result in the same error
```
$ python -m pl_examples.basic_examples.gpu_template --gpus 4 --distributed_backend ddp_spawn
$ python -m pl_examples.basic_examples.gpu_template --gpus 4 --distributed_backend ddp
```
```
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
CUDA_VISIBLE_DEVICES: [0,1,2,3]
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.6/site-packages/pl_examples/basic_examples/gpu_template.py", line 80, in <module>
    main(hyperparams)
  File "/opt/conda/lib/python3.6/site-packages/pl_examples/basic_examples/gpu_template.py", line 41, in main
    trainer.fit(model)
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 860, in fit
    self.barrier('fit_prepare_data')
  File "/opt/conda/lib/python3.6/site-packages/pytorch_lightning/trainer/trainer.py", line 1261, in barrier
    torch_distrib.barrier()
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 1484, in barrier
    _check_default_pg()
  File "/opt/conda/lib/python3.6/site-packages/torch/distributed/distributed_c10d.py", line 187, in _check_default_pg
    "Default process group is not initialized"
AssertionError: Default process group is not initialized
```

--- a/docs/source/trainer.rst
+++ b/docs/source/trainer.rst
@@ -9,7 +9,7 @@ Trainer
    :exclude-members:
         run_pretrain_routine,
         _abc_impl,
-        _Trainer_set_random_port,
+        set_random_port,
         _Trainer__set_root_gpu,
         _Trainer__init_optimizers,
         _Trainer__parse_gpu_ids,

--- a/pytorch_lightning/core/hooks.py
+++ b/pytorch_lightning/core/hooks.py
@@ -22,7 +22,7 @@ class ModelHooks(Module):
         Called at the beginning of fit and test.
 
         Args:
-            stage: either 'fit' or 'test'
+            step: either 'fit' or 'test'
         """
 
     def teardown(self, stage: str):

--- a/pytorch_lightning/core/hooks.py
+++ b/pytorch_lightning/core/hooks.py
@@ -22,7 +22,7 @@ class ModelHooks(Module):
         Called at the beginning of fit and test.
 
         Args:
-            step: either 'fit' or 'test'
+            stage: either 'fit' or 'test'
         """
 
     def teardown(self, stage: str):
@@ -30,7 +30,7 @@ class ModelHooks(Module):
         Called at the end of fit and test.
 
         Args:
-            step: either 'fit' or 'test'
+            stage: either 'fit' or 'test'
         """
 
     def on_fit_start(self):

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -475,6 +475,11 @@ class TrainerDDPMixin(ABC):
         model.trainer = self
         model.init_ddp_connection(self.global_rank, self.world_size, self.is_slurm_managing_tasks)
 
+        # call setup after the ddp process has connected
+        self.setup()
+        if self.is_function_implemented('setup'):
+            model.setup()
+
         # on world_size=0 let everyone know training is starting
         if self.is_global_zero:
             log.info('-' * 100)

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -185,6 +185,10 @@ class TrainerDDPMixin(ABC):
     tpu_cores: int
 
     @property
+    def is_function_implemented(self, *args) -> bool:
+        """Warning: this is just empty shell for code implemented in other class."""
+
+    @property
     def is_global_zero(self) -> int:
         """Warning: this is just empty shell for code implemented in other class."""
 

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -477,7 +477,7 @@ class TrainerDDPMixin(ABC):
 
         # call setup after the ddp process has connected
         self.setup()
-        if self.is_function_implemented('setup'):
+        if self.is_function_implemented('setup', model):
             model.setup()
 
         # on world_size=0 let everyone know training is starting

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -185,10 +185,6 @@ class TrainerDDPMixin(ABC):
     tpu_cores: int
 
     @property
-    def is_function_implemented(self, *args):
-        """Warning: this is just empty shell for code implemented in other class."""
-
-    @property
     def is_global_zero(self) -> int:
         """Warning: this is just empty shell for code implemented in other class."""
 

--- a/pytorch_lightning/trainer/distrib_data_parallel.py
+++ b/pytorch_lightning/trainer/distrib_data_parallel.py
@@ -185,7 +185,7 @@ class TrainerDDPMixin(ABC):
     tpu_cores: int
 
     @property
-    def is_function_implemented(self, *args) -> bool:
+    def is_function_implemented(self, *args):
         """Warning: this is just empty shell for code implemented in other class."""
 
     @property

--- a/pytorch_lightning/trainer/distrib_parts.py
+++ b/pytorch_lightning/trainer/distrib_parts.py
@@ -155,6 +155,11 @@ class TrainerDPMixin(ABC):
         return move_data_to_device(batch, device)
 
     def single_gpu_train(self, model):
+        # call setup
+        self.setup('fit')
+        if self.is_function_implemented('setup', model):
+            model.setup('fit')
+
         model.cuda(self.root_gpu)
 
         # CHOOSE OPTIMIZER
@@ -171,6 +176,11 @@ class TrainerDPMixin(ABC):
         self.run_pretrain_routine(model)
 
     def tpu_train(self, tpu_core_idx, model):
+        # call setup after the ddp process has connected
+        self.setup('fit')
+        if self.is_function_implemented('setup', model):
+            model.setup('fit')
+
         # put model on tpu
         self._device = xm.xla_device(self.tpu_id) if self.tpu_id is not None else xm.xla_device()
         model.to(self._device)
@@ -205,6 +215,10 @@ class TrainerDPMixin(ABC):
             self.save_spawn_weights(model)
 
     def dp_train(self, model):
+        # call setup after the ddp process has connected
+        self.setup('fit')
+        if self.is_function_implemented('setup', model):
+            model.setup('fit')
 
         # CHOOSE OPTIMIZER
         # allow for lr schedulers as well
@@ -246,6 +260,11 @@ class TrainerDPMixin(ABC):
         model.forward = model_autocast_original_forward
 
     def horovod_train(self, model):
+        # call setup after the ddp process has connected
+        self.setup()
+        if self.is_function_implemented('setup', model):
+            model.setup()
+
         if torch.cuda.is_available() and self.on_gpu:
             # Horovod: pin GPU to local rank
             assert self.root_gpu == hvd.local_rank()

--- a/pytorch_lightning/trainer/distrib_parts.py
+++ b/pytorch_lightning/trainer/distrib_parts.py
@@ -261,9 +261,9 @@ class TrainerDPMixin(ABC):
 
     def horovod_train(self, model):
         # call setup after the ddp process has connected
-        self.setup()
+        self.setup('fit')
         if self.is_function_implemented('setup', model):
-            model.setup()
+            model.setup('fit')
 
         if torch.cuda.is_available() and self.on_gpu:
             # Horovod: pin GPU to local rank

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1256,14 +1256,6 @@ class Trainer(
                 raise MisconfigurationException('You have defined `test_step()` but did not'
                                                 ' implement `test_dataloader` nor passed in `.test(test_dataloader)`.')
 
-    def barrier(self, name):
-        if self.use_ddp or self.use_ddp2:
-            torch_distrib.barrier()
-
-        if self.on_tpu and XLA_AVAILABLE:
-            # wait for all processes to catch up
-            torch_xla.core.xla_model.rendezvous(f'pl.Trainer.{name}')
-
 
 class _PatchDataLoader(object):
     r"""

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -857,8 +857,6 @@ class Trainer(
             model.prepare_data()
             self._is_data_prepared = True
 
-        # no barrier needed because all non-download processes will wait at init distributed (the barrier)
-
         self.setup('fit')
         if self.is_function_implemented('setup', model):
             model.setup('fit')

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -857,10 +857,6 @@ class Trainer(
             model.prepare_data()
             self._is_data_prepared = True
 
-        self.setup('fit')
-        if self.is_function_implemented('setup', model):
-            model.setup('fit')
-
         # Run auto batch size scaling
         if self.auto_scale_batch_size:
             if isinstance(self.auto_scale_batch_size, bool):
@@ -895,19 +891,19 @@ class Trainer(
                 self.ddp_train(task, model)
 
             elif self.distributed_backend == 'cpu_ddp':
-                self._set_random_port
+                self.set_random_port()
                 self.model = model
                 mp.spawn(self.ddp_train, nprocs=self.num_processes, args=(model,))
 
             elif self.distributed_backend == 'ddp_spawn':
-                self._set_random_port
+                self.set_random_port()
                 model.share_memory()
 
                 # spin up peers
                 mp.spawn(self.ddp_train, nprocs=self.num_processes, args=(model, ))
 
             elif self.distributed_backend == 'ddp':
-                self._set_random_port
+                self.set_random_port()
                 self.spawn_ddp_children(model)
 
         # 1 gpu or dp option triggers training using DP module
@@ -930,6 +926,9 @@ class Trainer(
             # track for predict
             self.model = model
 
+            # wait for all prepare data nodes to finish
+            self.barrier('setup')
+
             # train
             if self.tpu_id is not None:
                 self.tpu_train(self.tpu_id, model)
@@ -945,6 +944,11 @@ class Trainer(
             # run through amp wrapper
             if self.use_amp:
                 raise MisconfigurationException('amp + cpu is not supported.  Please use a GPU option')
+
+            # call setup after the ddp process has connected
+            self.setup('fit')
+            if self.is_function_implemented('setup', model):
+                model.setup('fit')
 
             # CHOOSE OPTIMIZER
             # allow for lr schedulers as well
@@ -1151,7 +1155,7 @@ class Trainer(
         if self.is_function_implemented('setup', model_ref):
             model_ref.setup('test')
 
-        # no barrier needed because all processes will catch up once the distributed group boots
+        self.barrier('test_setup')
 
         if model is None and ckpt_path == 'best' and self.checkpoint_callback.save_top_k <= 0:
             raise MisconfigurationException(
@@ -1253,6 +1257,14 @@ class Trainer(
             if self.testing and self.is_overridden('test_step', model):
                 raise MisconfigurationException('You have defined `test_step()` but did not'
                                                 ' implement `test_dataloader` nor passed in `.test(test_dataloader)`.')
+
+    def barrier(self, name):
+        if self.use_ddp or self.use_ddp2:
+            torch_distrib.barrier()
+
+        if self.on_tpu and XLA_AVAILABLE:
+            # wait for all processes to catch up
+            torch_xla.core.xla_model.rendezvous(f'pl.Trainer.{name}')
 
 
 class _PatchDataLoader(object):

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -857,7 +857,7 @@ class Trainer(
             model.prepare_data()
             self._is_data_prepared = True
 
-        self.barrier('fit_prepare_data')
+        # no barrier needed because all non-download processes will wait at init distributed (the barrier)
 
         self.setup('fit')
         if self.is_function_implemented('setup', model):
@@ -1153,7 +1153,7 @@ class Trainer(
         if self.is_function_implemented('setup', model_ref):
             model_ref.setup('test')
 
-        self.barrier('test_setup')
+        # no barrier needed because all processes will catch up once the distributed group boots
 
         if model is None and ckpt_path == 'best' and self.checkpoint_callback.save_top_k <= 0:
             raise MisconfigurationException(

--- a/pytorch_lightning/trainer/training_loop.py
+++ b/pytorch_lightning/trainer/training_loop.py
@@ -153,6 +153,7 @@ from typing import Union, List
 import numpy as np
 import torch
 from torch.utils.data import DataLoader
+import torch.distributed as torch_distrib
 
 from pytorch_lightning import _logger as log
 from pytorch_lightning.callbacks.base import Callback
@@ -701,6 +702,11 @@ class TrainerTrainLoopMixin(ABC):
     def run_training_teardown(self):
         if hasattr(self, '_teardown_already_run') and self._teardown_already_run:
             return
+
+        # clean up dist group
+        if self.use_ddp or self.use_ddp2:
+            torch_distrib.destroy_process_group()
+
         # Train end events
         with self.profiler.profile('on_train_end'):
             # callbacks

--- a/pytorch_lightning/trainer/training_loop.py
+++ b/pytorch_lightning/trainer/training_loop.py
@@ -259,7 +259,7 @@ class TrainerTrainLoopMixin(ABC):
         """Warning: this is just empty shell for code implemented in other class."""
 
     @abstractmethod
-    def is_function_implemented(self, *args):
+    def is_function_implemented(self, *args, **kwargs):
         """Warning: this is just empty shell for code implemented in other class."""
 
     @abstractmethod

